nixos/modules/services/torrent/transmission.nix

```diff
@@ -104,47 +99,79 @@ in
   };
 
   config = mkIf cfg.enable {
-    systemd.tmpfiles.rules = [
-      "d '${homeDir}' 0770 '${cfg.user}' '${cfg.group}' - -"
-      "d '${settingsDir}' 0700 '${cfg.user}' '${cfg.group}' - -"
-      "d '${fullSettings.download-dir}' '${downloadDirPermissions}' '${cfg.user}' '${cfg.group}' - -"
-      "d '${fullSettings.incomplete-dir}' '${downloadDirPermissions}' '${cfg.user}' '${cfg.group}' - -"
+    systemd.tmpfiles.rules =
+      optionals (! hasPrefix "/var/lib/" cfg.home) [
+        "d '${cfg.home}' 755 '${cfg.user}' '${cfg.group}' - -"
+        # Avoid "Unsafe path transition" by specifying parent directory .config
+        "d '${cfg.home}/.config' 755 '${cfg.user}' '${cfg.group}' - -"
+        "d '${cfg.home}/${settingsDir}' 700 '${cfg.user}' '${cfg.group}' - -"
+      ] ++
+      optionals (cfg.downloadDirPermissions != null) (
+        [ "d '${cfg.settings.download-dir}' '${cfg.downloadDirPermissions}' '${cfg.user}' '${cfg.group}' - -"
+        ] ++ optional cfg.settings.incomplete-dir-enabled
+          "d '${cfg.settings.incomplete-dir}' '${cfg.downloadDirPermissions}' '${cfg.user}' '${cfg.group}' - -"
+      );
+
+    assertions = [
+      { assertion = builtins.match "^/.*" cfg.home != null;
+        message = "`services.transmission.home' must be an absolute path.";
+      }
     ];
 
     systemd.services.transmission = {
       description = "Transmission BitTorrent Service";
       after = [ "network.target" ] ++ optional apparmor "apparmor.service";
       requires = mkIf apparmor [ "apparmor.service" ];
       wantedBy = [ "multi-user.target" ];
-
-      # 1) Only the "transmission" user and group have access to torrents.
-      # 2) Optionally update/force specific fields into the configuration file.
-      serviceConfig.ExecStartPre = preStart;
-      serviceConfig.ExecStart = "${pkgs.transmission}/bin/transmission-daemon -f --port ${toString config.services.transmission.port} --config-dir ${settingsDir}";
-      serviceConfig.ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
-      serviceConfig.User = cfg.user;
-      serviceConfig.Group = cfg.group;
-      # NOTE: transmission has an internal umask that also must be set (in settings.json)
-      serviceConfig.UMask = "0002";
+      preStart = ''
+        cp -f ${settingsFile} '${cfg.home}/${settingsDir}/settings.json'
+      '';
+
+      serviceConfig = {
+        WorkingDirectory = cfg.home;
+        ExecStart = "${pkgs.transmission}/bin/transmission-daemon -f --config-dir ${settingsDir}";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        User = cfg.user;
+        Group = cfg.group;
+        UMask = "0002";
```

Comment:
I don't know the implications of all the hardening options you add, and the umask as well, but here's how upstream's systemd service looks like:

```
[Unit]
Description=Transmission BitTorrent Daemon
After=network.target

[Service]
User=transmission
Type=notify
ExecStart=/usr/bin/transmission-daemon -f --log-error
ExecReload=/bin/kill -s HUP $MAINPID
NoNewPrivileges=true

[Install]
WantedBy=multi-user.target
```

I think we should stick to what upstream suggests.

P.S, our build of this package doesn't include this file, because we don't [do this](https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/transmission#n50).

Reply:
IMHO, hardening (especially network facing) services is important, and it's already done for a few services in Nixpkgs, like:
- `iodine`: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/iodine.nix#L140
- or `pppd`: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/pppd.nix#L90

Reply:
So, the current proposal now does like in `services.nginx`: set those sandboxing options only when `enableSandbox` is true.

Reply:
Awesome :rocket:. Thank you.